scripts/s3/migrate_to_external_accounts.py

```diff
@@ -0,0 +1,332 @@
+
+def make_new_node_settings(node, node_settings_document, external_account=None, user_settings_instance=None):
+    # kill backrefs to old models
+    database['node'].find_and_modify(
+        {'_id': node._id},
+        {
+            '$unset': {
+                '__backrefs.addons.addons3nodesettings': ''
+            }
+        }
+    )
+    node.reload()
+    node_settings_instance = node.get_or_add_addon('s3', auth=None, override=True, log=False)
+    node_settings_instance.bucket = node_settings_document['bucket']
+    node_settings_instance.save()
+    if external_account and user_settings_instance:
+        node_settings_instance.set_auth(
+            external_account,
+            user_settings_instance.owner,
+            log=False
+        )
+    return node_settings_instance
+
+def migrate(dry_run=True):
+    user_settings_list = list(database['addons3usersettings'].find())
+
+    # get in-memory versions of collections and collection sizes
+    old_user_settings_collection = database['addons3usersettings']
+    old_user_settings_count = old_user_settings_collection.count()
+    old_node_settings_collection = database['addons3nodesettings']
+    old_node_settings_count = old_node_settings_collection.count()
+
+    # Lists of IDs for logging purposes
+    external_accounts_created = []
+    migrated_user_settings = []
+    migrated_node_settings = []
+    deleted_user_settings = []
+    broken_user_settings = []
+    user_no_oauth_creds = []
+    invalid_oauth_creds = []
+    inactive_user_or_no_owner = []
+    unverifiable_node_settings = []
+    deleted_node_settings = []
+    nodeless_node_settings = []
+    duped_accounts = {}
+    dupe_count = 0
+
+    for user_settings_document in user_settings_list:
+        if user_settings_document['deleted']:
+            logger.info(
+                "Found addons3usersettings document (id:{0}) that is marked as deleted. It will not be migrated".format(user_settings_document['_id'])
+            )
+            deleted_user_settings.append(user_settings_document['_id'])
+            continue
+        if not verify_user_settings_documents(user_settings_document):
+            logger.info(
+                "Found broken addons3usersettings document (id:{0}) that could not be fixed.".format(user_settings_document['_id'])
+            )
+            broken_user_settings.append(user_settings_document['_id'])
+            continue
+        if not user_settings_document['access_key'] or not user_settings_document['secret_key']:
+            logger.info(
+                "Found addons3usersettings document (id:{0}) with incomplete or no oauth credentials. It will not be migrated.".format(user_settings_document['_id'])
+            )
+            user_no_oauth_creds.append(user_settings_document['_id'])
+            continue
+        external_account, user, new = migrate_to_external_account(user_settings_document)
+        if not external_account:
```

Comment:
It would be `None` when the oauth credentials are invalid, which prevents the creation of an `ExternalAccount` and causes `migrate_to_external_account` to [`return (None, None, None)`](https://github.com/CenterForOpenScience/osf.io/pull/4918/files#diff-f62665d47342aa3f7d5f293328733b4eR69)